From 8a4786ccf4a10254a9a759cf48b3cb472ef2ed1b Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:40:38 -0400
Subject: [ethernet] i40e: Add checks and message for Qualified Module info

Message-id: <1409658042-22689-278-git-send-email-sassmann@redhat.com>
Patchwork-id: 90470
O-Subject: [RHEL7.1 PATCH 277/281] i40e: Add checks and message for Qualified Module info
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Carolyn Wyborny <carolyn.wyborny@intel.com>

This patch adds a check during handle_link_event for unqualified
module when link is down and there is a module plugged.  If found,
print a message.

Change-ID: Ibd8666d77d3044c2a3dd4d762d3ae9ac6e18e943
Signed-off-by: Carolyn Wyborny <carolyn.wyborny@intel.com>
Tested-by: Jim Young <jamesx.m.young@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 7b592f6168e3091bc0868812b6a4a3126be5b82d)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_main.c b/drivers/net/ethernet/intel/i40e/i40e_main.c
index 5d26412..26c75d0 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_main.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_main.c
@@ -4480,6 +4480,13 @@ static int i40e_up_complete(struct i40e_vsi *vsi)
   netif_carrier_on(vsi->netdev);
  } else if (vsi->netdev) {
   i40e_print_link_message(vsi, false);
+  /* need to check for qualified module here*/
+  if ((pf->hw.phy.link_info.link_info &
+   I40E_AQ_MEDIA_AVAILABLE) &&
+      (!(pf->hw.phy.link_info.an_info &
+   I40E_AQ_QUALIFIED_MODULE)))
+   netdev_err(vsi->netdev,
+       "the driver failed to link because an unqualified module was detected.");
  }
 
  /* replay FDIR SB filters */
@@ -5492,6 +5499,13 @@ static void i40e_handle_link_event(struct i40e_pf *pf,
  memcpy(&pf->hw.phy.link_info_old, hw_link_info,
         sizeof(pf->hw.phy.link_info_old));
 
+ /* check for unqualified module, if link is down */
+ if ((status->link_info & I40E_AQ_MEDIA_AVAILABLE) &&
+     (!(status->an_info & I40E_AQ_QUALIFIED_MODULE)) &&
+     (!(status->link_info & I40E_AQ_LINK_UP)))
+  dev_err(&pf->pdev->dev,
+   "The driver failed to link because an unqualified module was detected.\n");
+
  /* update link status */
  hw_link_info->phy_type = (enum i40e_aq_phy_type)status->phy_type;
  hw_link_info->link_speed = (enum i40e_aq_link_speed)status->link_speed;
-- 
1.7.1